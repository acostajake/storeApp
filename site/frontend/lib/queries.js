import gql from 'graphql-tag';

import { perPage } from '../config';

export const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$desc: String!
		$image: String
		$largeImg: String
		$price: Int!
	) {
		createItem(
			title: $title
			desc: $desc
			image: $image
			largeImg: $largeImg
			price: $price
		) {
			id
		}
	}
`;

export const DELETE_ITEM_MUTATION = gql`
	mutation DELETE_ITEM_MUTATION($id: ID!) {
		deleteItem(id: $id) {
			id
		}
	}
`;

export const UPDATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$id: ID!
		$title: String
		$desc: String
		$price: Int
	) {
		updateItem(id: $id, title: $title, desc: $desc, price: $price) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ALL_ITEMS = gql`
	query GET_ALL_ITEMS($skip: Int = 0, $first: Int = ${perPage}) {
		items(first: $first, skip: $skip, orderBy: createdAt_DESC) {
			id
			title
			desc
			image
			largeImg
			price
		}
	}
`;

export const GET_ITEM = gql`
	query GET_ITEM($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			price
		}
	}
`;

export const GET_ITEM_WITH_IMAGE = gql`
	query GET_ITEM_WITH_IMAGE($id: ID!) {
		item(where: { id: $id }) {
			id
			title
			desc
			largeImg
		}
	}
`;

export const PAGINATION_QUERY = gql`
	query PAGINATION_QUERY {
		itemsConnection {
			aggregate {
				count
			}
		}
	}
`;

// Auth
export const RESET_MUTATION = gql`
	mutation RESET_MUTATION(
		$resetToken: String!
		$password: String!
		$confirmPassword: String!
	) {
		resetPassword(
			resetToken: $resetToken
			password: $password
			confirmPassword: $confirmPassword
		) {
			id
			name
			email
		}
	}
`;

export const REQUEST_RESET_MUTATION = gql`
	mutation REQUEST_RESET_MUTATION($email: String!) {
		requestReset(email: $email) {
			message
		}
	}
`;

export const SIGNUP_MUTATION = gql`
	mutation SIGNUP_MUTATION(
		$name: String!
		$email: String!
		$password: String!
	) {
		signup(name: $name, email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNIN_MUTATION = gql`
	mutation SIGNIN_MUTATION($email: String!, $password: String!) {
		signin(email: $email, password: $password) {
			id
			name
			email
		}
	}
`;

export const SIGNOUT_MUTATION = gql`
	mutation SIGNOUT_MUTATION {
		signout {
			message
		}
	}
`;

export const UPDATE_PERMISSIONS_MUTATION = gql`
	mutation updatePermissions($permissions: [Permission], $userId: ID!) {
		updatePermissions(permissions: $permissions, userId: $userId) {
			id
			permissions
			name
			email
		}
	}
`;

export const CURRENT_USER = gql`
	query {
		me {
			id
			email
			name
			permissions
		}
	}
`;

export const GET_ALL_USERS = gql`
	query {
		users {
			id
			name
			email
			permissions
		}
	}
`;
