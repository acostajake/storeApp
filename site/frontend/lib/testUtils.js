import casual from 'casual';
import {
	CURRENT_USER,
	PAGINATION_QUERY,
	REQUEST_RESET_MUTATION,
} from './queries';

// seed it so we get consistent results
casual.seed(777);

const fakeItem = () => ({
	__typename: 'Item',
	id: 'abc123',
	price: 5000,
	user: null,
	image: 'dog-small.jpg',
	title: 'dogs are best',
	desc: 'dogs',
	largeImg: 'dog.jpg',
});

const fakeUser = () => ({
	__typename: 'User',
	id: '4234',
	name: casual.name,
	email: casual.email,
	permissions: ['ADMIN'],
	orders: [],
	cart: [],
});

const fakeOrderItem = () => ({
	__typename: 'OrderItem',
	id: casual.uuid,
	image: `${casual.word}.jpg`,
	title: casual.words(),
	price: 4234,
	quantity: 1,
	desc: casual.words(),
});

const fakeOrder = () => ({
	__typename: 'Order',
	id: 'ord123',
	charge: 'ch_123',
	total: 40000,
	items: [fakeOrderItem(), fakeOrderItem()],
	createdAt: '2018-04 - 06T19: 24: 16.000Z',
	user: fakeUser(),
});

const fakeCartItem = (overrides) => ({
	__typename: 'CartItem',
	id: 'omg123',
	quantity: 3,
	item: fakeItem(),
	user: fakeUser(),
	...overrides,
});

// Fake LocalStorage
class LocalStorageMock {
	constructor() {
		this.store = {};
	}

	clear() {
		this.store = {};
	}

	getItem(key) {
		return this.store[key] || null;
	}

	setItem(key, value) {
		this.store[key] = value.toString();
	}

	removeItem(key) {
		delete this.store[key];
	}
}

const notSignedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: null } },
	},
];

const signedInMocks = [
	{
		request: { query: CURRENT_USER },
		result: { data: { me: fakeUser() } },
	},
];

const signedInWithCartMocks = [
	{
		request: { query: CURRENT_USER },
		result: {
			data: {
				me: {
					...fakeUser(),
					cart: [fakeCartItem(), fakeCartItem(), fakeCartItem()],
				},
			},
		},
	},
];

const makeMocksFor = (length) => {
	return [
		{
			request: { query: PAGINATION_QUERY },
			result: {
				data: {
					itemsConnection: {
						__typename: 'aggregate',
						aggregate: {
							__typename: 'count',
							count: length,
						},
					},
				},
			},
		},
	];
};

const requestResetMock = [
	{
		request: {
			query: REQUEST_RESET_MUTATION,
			variables: { email: 'test@fakeemail.com' },
		},
		result: {
			data: {
				requestReset: {
					message: 'success',
					__typename: 'Message',
				},
			},
		},
	},
];

export {
	LocalStorageMock,
	fakeItem,
	fakeUser,
	fakeCartItem,
	fakeOrder,
	fakeOrderItem,
	notSignedInMocks,
	requestResetMock,
	signedInMocks,
	signedInWithCartMocks,
	makeMocksFor,
};
