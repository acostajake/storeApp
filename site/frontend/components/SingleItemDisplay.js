import Head from 'next/head';
import React from 'react';
import { Query } from 'react-apollo';
import styled from 'styled-components';

import { GET_ITEM_WITH_IMAGE } from '../lib/queries';
import ErrorMessage from './ErrorMessage';

const SingleStyle = styled.div`
	max-width: 1200px;
	margin: 2rem 0 auto;
	box-shadow: ${(props) => props.theme.bs};
	display: grid;
	grid-auto-columns: 1fr;
	grid-auto-flow: column;
	min-height: 800px;
	img {
		width: 100%;
		height: 100%;
		object-fit: contain;
	}
	.details {
		margin: 3rem;
		font-size: 2rem;
	}
`;

const SingleItemDisplay = (props) => {
	return (
		<Query
			query={GET_ITEM_WITH_IMAGE}
			variables={{
				id: props.id,
			}}
		>
			{({ data, error, loading }) => {
				if (error) return <ErrorMessage error={error} />;
				if (loading) return <p>Loading item...</p>;
				const item = data.item;
				if (!item) return <p>Nothing found for {props.id}</p>;
				return (
					<SingleStyle>
						<Head>
							<title>Store | {item.title}</title>
						</Head>
						<img src={item.largeImg} alt={`${item.title} image`} />
						<div>
							<h2 className='details'>Viewing {item.title}</h2>
							<p>{item.desc}</p>
						</div>
					</SingleStyle>
				);
			}}
		</Query>
	);
};

export default SingleItemDisplay;
