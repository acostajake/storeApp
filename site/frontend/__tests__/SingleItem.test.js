import { mount } from 'enzyme';
import toJSON from 'enzyme-to-json';
import { MockedProvider } from 'react-apollo/test-utils';
import wait from 'waait';
import SingleItemDisplay from '../components/SingleItemDisplay';
import { GET_ITEM_WITH_IMAGE } from '../lib/queries';
import { fakeItem } from '../lib/testUtils';

const mocks = [
	{
		request: { query: GET_ITEM_WITH_IMAGE, variables: { id: '123' } },
		result: { data: { item: fakeItem() } },
	},
];

describe('<SingleItemDisplay/>', () => {
	it('renders with loading state', async () => {
		const mountItem = mount(
			<MockedProvider mocks={mocks}>
				<SingleItemDisplay id='123' />
			</MockedProvider>
		);
		expect(mountItem.text()).toContain('Loading item...');
	});
	it('renders with data', async () => {
		const mountItem = mount(
			<MockedProvider mocks={mocks}>
				<SingleItemDisplay id='123' />
			</MockedProvider>
		);
		await wait();
		mountItem.update();
		expect(toJSON(mountItem.find('h2'))).toMatchSnapshot();
		expect(toJSON(mountItem.find('img'))).toMatchSnapshot();
		expect(toJSON(mountItem.find('p'))).toMatchSnapshot();
	});
	it('errors with unfound item id', async () => {
		const errorMsg = {};
		const badMocks = [
			{
				request: { query: GET_ITEM_WITH_IMAGE, variables: { id: '123' } },
				result: { errors: [{ message: 'Item not found' }] },
			},
		];
		const mountItem = mount(
			<MockedProvider mocks={badMocks}>
				<SingleItemDisplay id='123' />
			</MockedProvider>
		);
		await wait();
		mountItem.update();
		const itemError = mountItem.find('[data-test="graphql-error"]');
		expect(toJSON(itemError)).toMatchSnapshot();
		expect(itemError.text()).toContain('Item not found');
	});
});
