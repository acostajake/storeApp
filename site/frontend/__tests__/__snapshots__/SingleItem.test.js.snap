// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SingleItemDisplay/> errors with unfound item id 1`] = `
<p
  data-test="graphql-error"
>
  <strong>
    Shoot!
  </strong>
  Item not found
</p>
`;

exports[`<SingleItemDisplay/> renders with data 1`] = `
<h2
  className="details"
>
  Viewing 
  dogs are best
</h2>
`;

exports[`<SingleItemDisplay/> renders with data 2`] = `
<img
  alt="dogs are best image"
  src="dog.jpg"
/>
`;

exports[`<SingleItemDisplay/> renders with data 3`] = `
<p>
  dogs
</p>
`;
